File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/FindElements_1261.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

import java.util.HashSet;

public class FindElements_1261 {
    HashSet<Integer> set;

    public FindElements_1261(TreeNode root) {
        set = new HashSet<>();
        helper(root, 0);
    }

    private void helper(TreeNode root, int val) {
        if (root == null) return;
        root.val = val;
        set.add(val);
        helper(root.left, val * 2 + 1);
        helper(root.right, val * 2 + 2);
    }

    public boolean find(int target) {
        return set.contains(target);
    }
}

File: Clion/zero/1400_1600/FindElements_1261.cpp
//
// Created by TzashiNorpu on 2023/6/21.
//
#include "../header/header.h"

using namespace zero;

FindElements::FindElements(TreeNode *root) {
  this->root = root;
  function<void(TreeNode *, int)> helper = [&](TreeNode *node, int value) -> void {
      if (!node) return;
      node->val = value;
      this->values.insert(value);
      helper(node->left, 2 * value + 1);
      helper(node->right, 2 * value + 2);
  };
  helper(root, 0);
}

bool FindElements::find(int target) {
  return this->values.find(target) != this->values.end();
}